crates/bevy_ecs/src/schedule_v3/schedule.rs

```diff
@@ -0,0 +1,1027 @@
+            for set in ambiguous_with.iter() {
+                if !self.system_set_ids.contains_key(set) {
+                    self.add_set(set.dyn_clone());
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn update_graphs(&mut self, id: NodeId, graph_info: GraphInfo) -> Result<(), BuildError> {
+        self.check_sets(&id, &graph_info)?;
+        self.check_edges(&id, &graph_info)?;
+        self.changed = true;
+
+        let GraphInfo {
+            sets,
+            dependencies,
+            ambiguous_with,
+        } = graph_info;
+
+        if !self.hierarchy.graph.contains_node(id) {
+            self.hierarchy.graph.add_node(id);
+        }
+
+        for set in sets.into_iter().map(|set| self.system_set_ids[&set]) {
+            self.hierarchy.graph.add_edge(set, id, ());
+        }
+
+        if !self.dependency.graph.contains_node(id) {
+            self.dependency.graph.add_node(id);
+        }
+
+        for (edge_kind, set) in dependencies
+            .into_iter()
+            .map(|(edge_kind, set)| (edge_kind, self.system_set_ids[&set]))
+        {
+            let (lhs, rhs) = match edge_kind {
+                DependencyEdgeKind::Before => (id, set),
+                DependencyEdgeKind::After => (set, id),
+            };
+            self.dependency.graph.add_edge(lhs, rhs, ());
+        }
+
+        match ambiguous_with {
+            Ambiguity::Check => (),
+            Ambiguity::IgnoreWithSet(ambigous_with) => {
+                for set in ambigous_with
+                    .into_iter()
+                    .map(|set| self.system_set_ids[&set])
+                {
+                    self.ambiguous_with.add_edge(id, set, ());
+                }
+            }
+            Ambiguity::IgnoreAll => {
+                self.ambiguous_with_all.insert(id);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn initialize(&mut self, world: &mut World) {
+        for (id, uninit) in self.uninit.drain(..) {
+            match uninit {
+                UninitNode::System(mut system, mut conditions) => {
+                    debug_assert!(id.is_system());
+                    system.initialize(world);
+                    self.systems.insert(id, system);
+                    for condition in &mut conditions {
+                        condition.initialize(world);
+                    }
+                    self.conditions.insert(id, conditions);
+                }
+                UninitNode::SystemSet(mut conditions) => {
+                    debug_assert!(id.is_set());
+                    for condition in &mut conditions {
+                        condition.initialize(world);
+                    }
+                    self.conditions
+                        .entry(id)
+                        .or_insert_with(Vec::new)
+                        .append(&mut conditions);
+                }
+            }
+        }
+    }
+
+    fn build_dependency_graph(&mut self) -> Result<(), BuildError> {
```

Comment:
> Ex: we do all of the check_graph work for the hierarchy graph, then we do it again for the final flattened graph.

The `check_graph` for the hierarchy and flattened dependency graph accomplish different goals since the two graphs are "perpendicular".

When I wrote this, I was convinced it was all necessary to enforce the invariants. Some cyclic dependencies only appear once the dependency graph is flattened, so the hierarchy and initial dependency graphs passing does not guarantee the flattened graph will too.

Reply:
Sorry for the "pretty much identical" back to back comment threads. It looked like the Github vscode/pull requests review dropped one of these comments so I thought I had to re-type it. Apparently it didn't actually drop it :)